Prometheus metric for repository's last update

oxzi opened the pull request:
What is the purpose of this change? What does it change?
--------------------------------------------------------

A new gauge metric `rest_server_repo_last_update_timestamp` was added to monitor each repository's last write access. This allows a basic monitoring for each repository's freshness.


Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

This might be related to the feature request in #176.

Checklist
---------

- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/rest-server/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/rest-server/commits/master)
- [x] I'm done, this Pull Request is ready for review

wojas reviewed (commented):
I like having this feature, but I have concerns about the way it does its own repo scanning on startup. It is also unclear what the metric should show after the initial load.

Issue #176 so far only discusses metrics for disk size, not the last modified timestamp.

oxzi commented on handlers.go:
```diff
@@ -46,6 +49,93 @@ func httpDefaultError(w http.ResponseWriter, code int) {
+
+	var repoPaths []string
+
+	walkFunc := func(path string, d fs.DirEntry, err error) error {
```
The `folderPathValid` function will be used below in line 122, as both the repository's absolute path as well as the relative folder path are needed. I don't see any advantage in using the `splitURLPath` function directly, especially as both `repo.ObjectTypes` and `repo.FileTypes` are being checked in the `walkFunc` to determine if some directory path should be followed further.

As any occurrence of an `repo.ObjectTypes` like directory aborts this path by returning `filepath.SkipDir`, I don't expect too much directory walking. However, there might be special cases where this might perform poorly. I'm open to suggestions how to address this.

oxzi commented on cmd/rest-server/main.go:
```diff
@@ -126,6 +126,12 @@ func runRoot(cmd *cobra.Command, args []string) error {
 		log.Println("Private repositories disabled")
 	}
 
+	if server.Prometheus {
```
Thanks. I have added the `--prometheus-no-preload` flag to disable preloading.

oxzi commented on repo/repo.go:
```diff
@@ -123,6 +125,21 @@ const (
+		log.Printf("%v.PreloadMetrics()", h)
+	}
+
+	stat, err := os.Lstat(h.getSubPath("index"))
```
Changed!

oxzi commented on repo/repo.go:
```diff
@@ -113,6 +113,8 @@ const (
 	BlobWrite  = 'W' // A blob has been written
 	BlobDelete = 'D' // A blob has been deleted
+
+	RepoLastUpdate = 'U' // Set an repository's last update timestamp for preloading
```
The `RepoPreloadLastUpdate` flag - previously named `RepoLastUpdate` - is only used for preloading. That's also why I changed the name.
Later on every `BlobWrite`, the timestamp will also being bumped to the current time.

oxzi commented:
@wojas: Thanks a lot for your review! I have made some changes and answered in the conversations above.

Furthermore, I tried to motivate why I introduced the preloading, which is not directly part of those metrics, but are imho necessary for a complete view after (automatically) restarting the rest-server.

Please feel free to take another look; and thanks again!

riotbib commented:
This is relevant to me, please consider merging. Thanks!

MarcWort commented:
Any updates on this change? This would be a really useful feature.
@oxzi Would you mind rebasing this PR?

oxzi commented:
@MarcWort: Thanks for bumping this PR. Since I haven't received any feedback from @wojas after addressing the review and no other maintainer stepped up, I have kind of forgotten about this PR. For a while, I used my build of the rest-server but stopped using the rest-server completely in favor for another backend. If there is really interest in getting this merged, I would take another look and rebase it.

darkphoenix commented:
@oxzi: I would also very much appreciate this feature, it would be very useful to be able to ensure with monitoring that backups of clients are running regularly. Is there anything we could potentially do to help?

oxzi commented:
@darkphoenix: Thanks for the kind words. I would try to take a closer look at this PR this week and try to make it mergeable again.

However, the real "blocker" would be a decision by the project maintainers if they want to accept this PR. If they do not decide to do so, this would be in vain.
